### PR TITLE
[IOTDB-937] change jdbc into session and don't need to create metadata

### PR DESCRIPTION
**Pull request by Genius-pig**
https://issues.apache.org/jira/browse/IOTDB-937

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -347,14 +348,17 @@ CREATE TIMESERIES root.fit.d2.s3 WITH DATATYPE=INT32,ENCODING=RLE;
...
 ```
 Time,root.fit.d1.s1,root.fit.d1.s2,root.fit.d2.s1,root.fit.d2.s3,root.fit.p.s1
-1,100,'hello',200,300,400
-2,500,'world',600,700,800
-3,900,'IoTDB',1000,1100,1200
+1,100,hello,200,300,400
+2,500,world,600,700,800
+3,900,IoTDB,1000,1100,1200
 ```
+> if you have a comma in your field, the field should be enclosed by a double quote or single quote.
```
Comment (SteveYurongSu) [inline README.md:361]: As a user, I may have the following concerns:

1. What delimiters are allowed?
2. What are the special characters (e.g., `"` and `,`) that will affect data import? How should these special characters be handled?

I think the documentation is not clear enough.

**Review by SteveYurongSu — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (Genius-pig) [reply README.md]: updated

**Comment by Genius-pig** [comment]
@SteveYurongSu add tests, take a look.

File: README.md [hunk 2]
```diff
@@ -347,14 +348,17 @@ CREATE TIMESERIES root.fit.d2.s3 WITH DATATYPE=INT32,ENCODING=RLE;
...
 ### An example of import csv file
 
 ```
 Time,root.fit.d1.s1,root.fit.d1.s2,root.fit.d2.s1,root.fit.d2.s3,root.fit.p.s1
-1,100,'hello',200,300,400
-2,500,'world',600,700,800
-3,900,'IoTDB',1000,1100,1200
+1,100,hello,200,300,400
+2,500,world,600,700,800
+3,900,"hello, world",1000,1100,1200
```
Comment (SteveYurongSu) [inline README.md:359]:
```suggestion
3,900,"hello, \"world\"",1000,1100,1200
```

File: README.md [hunk 3]
```diff
@@ -347,14 +348,17 @@ CREATE TIMESERIES root.fit.d2.s3 WITH DATATYPE=INT32,ENCODING=RLE;
...
 ```
+> if you have some special characters like comma in your field, the field should be enclosed by a double quote or single quote. for example, "hello,\"great";
```
Comment (SteveYurongSu) [inline README.md:361]: Note that the following special characters in fields need to be checked before importing:
1. `,` : fields containing `,` should be quoted by a pair of `"` or a pair of `'`.
2. `"` : `"` in fields should be replaced by `\"` .

File: README_ZH.md [hunk 4]
```diff
@@ -347,16 +347,19 @@ CREATE TIMESERIES root.fit.d2.s1 WITH DATATYPE=INT32,ENCODING=RLE;
...
 ### 从 csv 文件导入数据的示例
 
 ```
 Time,root.fit.d1.s1,root.fit.d1.s2,root.fit.d2.s1,root.fit.d2.s3,root.fit.p.s1
-1,100,'hello',200,300,400
-2,500,'world',600,700,800
-3,900,'IoTDB',1000,1100,1200
+1,100,hello,200,300,400
+2,500,world,600,700,800
+3,900,IoTDB,1000,1100,1200
```
Comment (SteveYurongSu) [inline README_ZH.md:358]:
```suggestion
3,900,"hello, \"world\"",1000,1100,1200
```

File: README_ZH.md [hunk 5]
```diff
@@ -347,16 +347,19 @@ CREATE TIMESERIES root.fit.d2.s1 WITH DATATYPE=INT32,ENCODING=RLE;
...
 ```
 
+> 如果字段中有一些特殊字符，如逗号，字段应该用双引号或单引号括起来。例如, "hello,\"great";
```
Comment (SteveYurongSu) [inline README_ZH.md:361]: 注意，在导入数据前，需要特殊处理下列的字符：
1. `,` : 包含`,`的字段需要使用单引号或者双引号括起来
2. `"` : "字段中的`"`需要被替换成转义字符`\"`
3. `'` : "字段中的`'`需要被替换成转义字符`\'`

File: cli/src/main/java/org/apache/iotdb/tool/ExportCsv.java [hunk 6]
```diff
@@ -263,112 +228,104 @@ private static void dumpResult(String sql, int index)
...
-      if (i < count) {
-        bw.write(metadata.getColumnLabel(i) + ",");
-      } else {
-        bw.write(metadata.getColumnLabel(i) + "\n");
-      }
-      typeList.add(metadata.getColumnType(i));
+  private static void writeMetadata(BufferedWriter bw, List<String> columnNames)
+      throws IOException {
+    for (int i = 0; i < columnNames.size() - 1; i++) {
+      bw.write(columnNames.get(i) + ",");
```
Comment (SteveYurongSu) [inline cli/src/main/java/org/apache/iotdb/tool/ExportCsv.java:254]: How to deal with special characters in column names? This may cause problems when importing the CSV files generated by our export tool...

File: cli/src/test/java/org/apache/iotdb/tool/CsvLineSplitTest.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+package org.apache.iotdb.tool;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+public class CsvLineSplitTest {
+
+  @Test
+  public void testSplit() {
+    Assert.assertArrayEquals(new String[]{"", "a", "b", "c", "\\\""}, ImportCsv.splitCsvLine(",a,b,c,\"\\\"\""));
```
Comment (SteveYurongSu) [inline cli/src/test/java/org/apache/iotdb/tool/CsvLineSplitTest.java:28]: Please indicate in the document that "\" is also a special character...

File: cli/src/main/java/org/apache/iotdb/tool/ImportCsv.java [hunk 8]
```diff
@@ -119,296 +88,78 @@ private static Options createOptions() {
...
+              values.add(cols[measurementAndPosition.getValue()]);
+            }
+          }
+          measurementsList.add(measurements);
+          valuesList.add(values);
+        }
       }
-
-      executeSqls(bw, tmp, startTime, file);
+      session.insertRecords(devices, times, measurementsList, valuesList);
```
Comment (SteveYurongSu) [inline cli/src/main/java/org/apache/iotdb/tool/ImportCsv.java:141]: I think there may be a problem when importing millions of records at a time...

**Comment by SteveYurongSu** [comment]
Please add more tests for:
1. exporting aggregation query results
2. importing / exporting query date sets with special characters

File: README.md (hunk 3 above)
Comment (Genius-pig) [reply README.md]: update

File: README_ZH.md (hunk 5 above)
Comment (Genius-pig) [reply README_ZH.md]: update

File: cli/src/test/java/org/apache/iotdb/tool/CsvLineSplitTest.java (hunk 7 above)
Comment (Genius-pig) [reply cli/src/test/java/org/apache/iotdb/tool/CsvLineSplitTest.java:28]: update

**Comment by Genius-pig** [comment]
> Please add more tests for:
> 
> 1. exporting aggregation query results
> 2. importing/exporting query date sets with special characters

added.

File: cross-tests/src/test/java/org/apache/iotdb/cross/tests/tools/importCsv/ExportCsvTestIT.java [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+    session.open();
+
+    String deviceId = "root.sg1.d1";
+    List<String> measurements = new ArrayList<>();
+    measurements.add("s1");
+    measurements.add("s2");
+
+    List<String> values = new ArrayList<>();
+    values.add("1.0");
+    values.add("abc\",aa");
```
Comment (SteveYurongSu) [inline cross-tests/src/test/java/org/apache/iotdb/cross/tests/tools/importCsv/ExportCsvTestIT.java:187]: Tests about `'`?

File: cli/src/main/java/org/apache/iotdb/tool/ImportCsv.java [hunk 10]
```diff
@@ -119,296 +89,90 @@ private static Options createOptions() {
...
+        for (Entry<String, List<Integer>> deviceToPositions : devicesToPositions
+            .entrySet()) {
+          String device = deviceToPositions.getKey();
+          devices.add(device);
+
+          times.add(Long.parseLong(cols[0]));
+
+          List<String> values = new ArrayList<>();
+          for (int position : deviceToPositions.getValue()) {
+            if (cols[position].equals("") && cols[position].equals("null")) {
```
Comment (SteveYurongSu) [inline cli/src/main/java/org/apache/iotdb/tool/ImportCsv.java:136]: This equation is always false and ... 
Does `""` represent `null`?

**Review by SteveYurongSu — CHANGES_REQUESTED** [review]
Please add some tests for importing and exporting `null` values...

File: cli/src/main/java/org/apache/iotdb/tool/ExportCsv.java [hunk 11]
```diff
@@ -263,112 +228,123 @@ private static void dumpResult(String sql, int index)
...
+  private static void writeValue(List<Field> fields, BufferedWriter bw) throws IOException {
+    for (int j = 0; j < fields.size() - 1; j++) {
+      String value = fields.get(j).getStringValue();
+      if ("null".equalsIgnoreCase(value)) {
+        bw.write(",");
+      } else {
+        if (fields.get(j).getDataType() == TSDataType.TEXT) {
+          int location = value.indexOf("\"");
+          if (location > -1) {
+            if (value.charAt(location - 1) != '\\') {
```
[the remark this answers is not in the document]
Comment (Genius-pig) [reply cli/src/main/java/org/apache/iotdb/tool/ExportCsv.java:314]: because we need to know before  `"`, whether has a "\".

File: cross-tests/src/test/java/org/apache/iotdb/cross/tests/tools/importCsv/ExportCsvTestIT.java (hunk 9 above)
Comment (Genius-pig) [reply cross-tests/src/test/java/org/apache/iotdb/cross/tests/tools/importCsv/ExportCsvTestIT.java:188]: add

File: cli/src/main/java/org/apache/iotdb/tool/ImportCsv.java (hunk 10 above)
Comment (Genius-pig) [reply cli/src/main/java/org/apache/iotdb/tool/ImportCsv.java]: remove, "" represent "";

**Review by SteveYurongSu — APPROVED** [review]
LGTM
